tap_pardot/streams.py

```diff
@@ -143,6 +145,114 @@ def get_params(self):
+        )
+
+    def _get_window_state(self):
+        """
+        Pardot's API seems to treat dates like 'YYYY-mm-DD HH:MM:SS' in the
+        user's time zone, and dates like 'YYYY-mm-DDTHH:MM:SSZ in UTC.
+
+        For consistency:
+        - All datetimes should be in string format, never parsed.
+        - The window_start and window_end times should all be in UTC.
+        - The sub_window should be in the user's timezone.
+        """
+        window_start = singer.get_bookmark(self.state, self.stream_name, 'window_start')
+        sub_window_end = singer.get_bookmark(self.state, self.stream_name, 'sub_window_end')
+        window_end = singer.get_bookmark(self.state, self.stream_name, 'window_end')
+
+        start_date = self.config.get('start_date')
+
+        window_start = max(window_start, start_date)
+        window_end = window_end
+
+        return window_start, sub_window_end, window_end
+
+    def pre_sync(self):
+        if singer.get_bookmark(self.state, self.stream_name, 'sub_window_end') is None:
+            if singer.get_bookmark(self.state, self.stream_name, 'window_start') is None:
+                singer.write_bookmark(self.state, self.stream_name, "window_start", self.config.get('start_date'))
+            if singer.get_bookmark(self.state, self.stream_name, 'window_end') is None:
+                singer.write_bookmark(self.state, self.stream_name, "window_end", utils.strftime(utils.now()))
+        singer.write_state(self.state)
+
+    def post_sync(self):
+        # Set window_start to current window_end
+        window_start = singer.get_bookmark(self.state, self.stream_name, "window_end")
+        singer.write_bookmark(self.state, self.stream_name, "window_start", window_start)
+        singer.clear_bookmark(self.state, self.stream_name, "window_end")
+        singer.bookmarks.clear_bookmark(self.state, self.stream_name, "sub_window_end")
+        singer.write_state(self.state)
+
+
+    def check_order(self, current_bookmark_value):
+        if self._last_bookmark_value is None:
+            self._last_bookmark_value = current_bookmark_value
+
+        if current_bookmark_value > self._last_bookmark_value:
+            raise Exception(
+                "Detected out of order data. Current bookmark value {} is greater than last bookmark value {}".format(
+                    current_bookmark_value, self._last_bookmark_value
+                )
+            )
+
+        self._last_bookmark_value = current_bookmark_value
+
+    def get_records(self, window_start, window_end):
+        """ Make one page request and update bookmarks, sync handles pagination based on result size. """
+        replication_key = self.replication_key[0]
+        sub_window_end = window_end
+        data = self.client.get(self.endpoint, **self.get_params(window_start, window_end))
+
+        if data["result"] is None or data["result"].get("total_results") == 0:
+            return []
+
+        records = data["result"][self.data_key]
```

Comment:
Should we use `get()` instead of `[]` on many of the calls below here? An error here or below would indicate the record is missing its replication key and we might want to call that out more explicitly?

Reply:
Yeah, I think that's a good point. We're being strict with the ordering, so might as well have that check in there too 👍